virtcontainers/clh.go

```diff
@@ -0,0 +1,1316 @@
+//
+//###########################################################################
+
+func (clh *cloudHypervisor) addDevice(devInfo interface{}, devType deviceType) error {
+	span, _ := clh.trace("addDevice")
+	defer span.Finish()
+
+	var err error
+
+	switch v := devInfo.(type) {
+	case Endpoint:
+		clh.Logger().WithField("function", "addDevice").Info(fmt.Sprintf("Adding Endpoint of type %v", v))
+		clh.cliBuilder.AddNet(CLINet{
+			device: v.Name(),
+			mac:    v.HardwareAddr(),
+		})
+
+	default:
+		clh.Logger().WithField("function", "addDevice").Warn(fmt.Sprintf("Add device of type %v is not supported.", v))
+	}
+
+	return err
+}
+
+func (clh *cloudHypervisor) clhAddVsock(vs types.VSock) error {
+	clh.Logger().WithField("function", "clhAddVsock").Info("Add vsock device contextID=", vs.ContextID)
+	return nil
+}
+
+func (clh *cloudHypervisor) Logger() *log.Entry {
+	return virtLog.WithField("subsystem", "cloudHypervisor")
+}
+
+func (clh *cloudHypervisor) capabilities() types.Capabilities {
+	span, _ := clh.trace("capabilities")
+	defer span.Finish()
+
+	clh.Logger().WithField("function", "capabilities").Info("get Capabilities")
+	var caps types.Capabilities
+	return caps
+
+}
+
+func (clh *cloudHypervisor) trace(name string) (opentracing.Span, context.Context) {
+
+	if clh.ctx == nil {
+		clh.Logger().WithField("type", "bug").Error("trace called before context set")
+		clh.ctx = context.Background()
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(clh.ctx, name)
+
+	span.SetTag("subsystem", "cloudHypervisor")
+	span.SetTag("type", "clh")
+
+	return span, ctx
+}
+
+func (clh *cloudHypervisor) terminate() (err error) {
+	span, _ := clh.trace("terminate")
+	defer span.Finish()
+
+	defer func() {
+		if err != nil {
+			clh.Logger().Info("Terminate Cloud Hypervisor failed")
+		} else {
+			clh.Logger().Info("Cloud Hypervisor stopped")
+			clh.state.PID = 0
+			clh.state.VirtiofsdPID = 0
+			clh.state.state = clhNotReady
+			clh.storeState()
+			clh.Logger().Debug("removing virtiofsd and vm sockets")
+			path, err := clh.virtioFsSocketPath(clh.id)
+			if err == nil {
+				rerr := os.Remove(path)
+				if rerr != nil {
+					clh.Logger().WithField("path", path).Warn("removing virtiofsd socket failed")
+				}
+			}
+			path, err = clh.vsockSocketPath(clh.id)
+			if err == nil {
+				rerr := os.Remove(path)
+				if rerr != nil {
+					clh.Logger().WithField("path", path).Warn("removing vm socket failed")
+				}
+			}
+		}
+	}()
+
+	pid := clh.state.PID
+	if pid == 0 {
+		clh.Logger().WithField("PID", pid).Info("Skipping kill cloud hypervisor. invalid pid")
+		return nil
+	}
+	clh.Logger().WithField("PID", pid).Info("Stopping Cloud Hypervisor")
+
+	// Check if VM process is running, in case it is not, let's
+	// return from here.
+	if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
+		return nil
+	}
+
+	// Send a SIGTERM to the VM process to try to stop it properly
+	if err = syscall.Kill(pid, syscall.SIGTERM); err != nil {
```

Comment:
Having two kills like this is potentially racy. I'd remove the `Kill(0)` and instead `return nil` if `Kill(SIGTERM)` fails with `ESRCH`.

Reply:
Tis can be handled with a follow-on PR since this idiom is used for the other hypervisor implementations too.